core/text/src/markdown/parser.rs

```diff
@@ -0,0 +1,280 @@
+use super::*;
+use crate::{Color, Style};
+use nom::branch::alt;
+use nom::bytes::complete::*;
+use nom::combinator::*;
+use nom::multi::*;
+use nom::IResult;
+
+#[derive(Debug, PartialEq)]
+pub enum Token {
+    Color(ColorToken),
+    Style(StyleToken),
+    Text(String),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct ColorToken {
+    pub color: Color,
+    pub rest: Vec<Token>,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct StyleToken {
+    pub style: Style,
+    pub rest: Vec<Token>,
+}
+
+fn token(t: LexToken) -> impl Fn(Tokens) -> IResult<Tokens, LexToken> {
+    move |input: Tokens| {
+        let (rest, tok) = take(1usize)(input)?;
+        if tok.tok[0] == t {
+            Ok((rest, tok.tok[0].clone()))
+        } else {
+            Err(nom::Err::Error((rest, nom::error::ErrorKind::Tag)))
+        }
+    }
+}
+
+fn word_or_space(input: Tokens) -> IResult<Tokens, LexToken> {
+    let (rest, tok) = take(1usize)(input)?;
+    match tok.tok[0] {
+        LexToken::Word(_) | LexToken::Space(_) => Ok((rest, tok.tok[0].clone())),
+        _ => Err(nom::Err::Error((rest, nom::error::ErrorKind::Tag))),
+    }
+}
+
+pub fn tokens_to_text(tokens: Vec<LexToken>) -> Token {
+    let mut s = String::new();
+
+    for tok in tokens {
+        match tok {
+            LexToken::Word(word) => s.push_str(&word),
+            LexToken::Space(space) => s.push_str(&space),
+            _ => unreachable!(),
+        }
+    }
+
+    Token::Text(s.trim().to_string())
+}
+
+pub fn has_lbrace(input: Tokens) -> isize {
+    for (i, tok) in input.tok.iter().enumerate() {
+        if let LexToken::LBrace = tok {
+            return i as isize;
+        }
+    }
+
+    -1isize
+}
+
+pub fn parse_control_word(i: Tokens) -> IResult<Tokens, Token> {
+    let (i, _) = token(LexToken::ControlWordStarter)(i)?;
+    let (i, next) = take(1usize)(i)?;
+
+    match &next.tok[0] {
+        LexToken::Word(s) => match s.as_str() {
+            "red" => {
```

Comment:
I'm sure you're already planning to fix this, but to eliminate duplicate code there should be some enum to determine the formatting specifier in the input to replace the duplicated match arms.